nemo_text_processing/text_normalization/vi/verbalizers/time.py

```diff
@@ -0,0 +1,149 @@
+# limitations under the License.
+
+import pynini
+from pynini.lib import pynutil
+
+from nemo_text_processing.text_normalization.vi.graph_utils import (
+    NEMO_NOT_QUOTE,
+    GraphFst,
+    convert_space,
+    delete_preserve_order,
+    delete_space,
+)
+from nemo_text_processing.text_normalization.vi.utils import get_abs_path
+
+
+class TimeFst(GraphFst):
+    """
+    Finite state transducer for verbalizing Vietnamese time.
+
+    Converts tagged time entities into spoken form, e.g.:
+    - time { hours: "tám" minutes: "ba mươi" } -> tám giờ ba mươi phút
+    - time { hours: "mười bốn" minutes: "mười lăm" } -> mười bốn giờ mười lăm phút
+    - time { hours: "chín" } -> chín giờ
+    - time { minutes: "ba" seconds: "hai mươi" } -> ba phút hai mươi giây
+    - time { hours: "tám" minutes: "hai mươi ba" zone: "g m t" } -> tám giờ hai mươi ba phút GMT
+
+    Args:
+        deterministic: if True will provide a single transduction option,
+            for False multiple transduction are generated (used for audio-based normalization)
+    """
+
+    def __init__(self, deterministic: bool = True):
+        super().__init__(name="time", kind="verbalize", deterministic=deterministic)
+
+        time_zone = convert_space(pynini.string_file(get_abs_path("data/time/time_zones.tsv")))
+        quoted_text = pynini.closure(NEMO_NOT_QUOTE)
+
+        def extract_field(field_name):
+            return (
+                pynutil.delete(f"{field_name}:")
+                + delete_space
+                + pynutil.delete("\"")
+                + quoted_text
+                + pynutil.delete("\"")
+            )
+
+        hour_component = extract_field("hours")
+        timezone_component = extract_field("zone") @ time_zone
+
+        zero_minute_component = pynutil.delete("minutes:") + delete_space + pynutil.delete("\"không\"")
+        zero_second_component = pynutil.delete("seconds:") + delete_space + pynutil.delete("\"không\"")
+
+        non_zero_minute_component = (
+            pynutil.delete("minutes:")
+            + delete_space
+            + pynutil.delete("\"")
+            + pynini.closure(NEMO_NOT_QUOTE - pynini.accep("không"))
+            + pynutil.delete("\"")
+        )
+        non_zero_second_component = (
+            pynutil.delete("seconds:")
+            + delete_space
+            + pynutil.delete("\"")
+            + pynini.closure(NEMO_NOT_QUOTE - pynini.accep("không"))
+            + pynutil.delete("\"")
+        )
+
+        hour_with_unit = hour_component + pynutil.insert(" giờ")
+        minute_with_unit = non_zero_minute_component + pynutil.insert(" phút")
+        second_with_unit = non_zero_second_component + pynutil.insert(" giây")
+
+        optional_timezone = pynini.closure(delete_space + pynutil.insert(" ") + timezone_component, 0, 1)
+        optional_preserve_order = pynini.closure(delete_space + delete_preserve_order, 0, 1)
+
+        patterns = [
+            hour_with_unit
```

Comment:
ditto regarding other comment. If it's multiline concatenations in a list, it becomes difficult to read on maintenance